src/issues/issueFeatureRegistrar.ts

```diff
@@ -0,0 +1,70 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { PullRequestManager } from '../github/pullRequestManager';
+import * as vscode from 'vscode';
+import { NewIssue } from './issueTodoProvider';
+import { IssueHoverProvider } from './issueHoverProvider';
+import * as LRUCache from 'lru-cache';
+import { IssueLinkProvider } from './issueLinkProvider';
+import { UserHoverProvider } from './userHoverProvider';
+import { IssueTodoProvider } from './issueTodoProvider';
+import { PullRequestModel } from '../github/pullRequestModel';
+
+export class IssueFeatureRegistrar implements vscode.Disposable {
+	constructor(context: vscode.ExtensionContext, private manager: PullRequestManager) {
+		const resolvedIssues: LRUCache<string, PullRequestModel> = new LRUCache(50); // 50 seems big enough
+		context.subscriptions.push(vscode.commands.registerCommand('issue.createIssueFromSelection', this.createTodoIssue, this));
+		context.subscriptions.push(vscode.languages.registerHoverProvider('*', new IssueHoverProvider(manager, resolvedIssues)));
+		context.subscriptions.push(vscode.languages.registerHoverProvider('*', new UserHoverProvider(manager)));
+		context.subscriptions.push(vscode.languages.registerDocumentLinkProvider('*', new IssueLinkProvider(manager, resolvedIssues)));
+		context.subscriptions.push(vscode.languages.registerCodeActionsProvider('*', new IssueTodoProvider()));
+	}
+
+	dispose() { }
+
+	async createTodoIssue(newIssue?: NewIssue) {
+		let document: vscode.TextDocument;
+		let titlePlaceholder: string | undefined;
+		let insertIndex: number | undefined;
+		let lineNumber: number | undefined;
+		let range: vscode.Range;
+		if (!newIssue && vscode.window.activeTextEditor) {
+			document = vscode.window.activeTextEditor.document;
+			range = vscode.window.activeTextEditor.selection;
+		} else if (newIssue) {
+			document = newIssue.document;
+			insertIndex = newIssue.insertIndex;
+			lineNumber = newIssue.lineNumber;
+			titlePlaceholder = newIssue.line.substring(insertIndex + 4, newIssue.line.length).trim();
+			range = newIssue.range;
+		} else {
+			return undefined;
+		}
+
+		const title = await vscode.window.showInputBox({ value: titlePlaceholder, prompt: 'Issue title' });
+		if (title) {
+			let codePermalink: string | undefined;
+			if (this.manager.repository.state.HEAD && this.manager.repository.state.HEAD.commit) {

```

Comment:
Not sure what the best way to handle this case is, but if I have a local HEAD commit that hasn't been pushed to the remote and create a TODO issue, the link will 404. There's an `ahead` property on `HEAD` that could be used to check this

Reply:
Good catch! I'm adding a fallback to just use a link, not a permalink in this case. You could still hit an issue where the file doesn't exist on the remote, but it should be less common.